src/c2ffi/c2ffi.lisp

```diff
@@ -0,0 +1,163 @@
+;;; EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+;;; MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+;;; NONINFRINGEMENT.  IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+;;; HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+;;; WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+;;; OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+;;; DEALINGS IN THE SOFTWARE.
+;;;
+
+(in-package #:cffi/c2ffi)
+
+;;; NOTE: Most of this has been taken over from cl-autowrap.
+
+;;; Note this is rather untested and not very extensive at the moment;
+;;; it should probably work on linux/win/osx though. Patches welcome.
+
+(defun local-cpu ()
+  #+x86-64 "x86_64"
+  #+(and (not (or x86-64 freebsd)) x86) "i686"
+  #+(and (not x86-64) x86 freebsd) "i386"
+  #+arm "arm")
+
+(defun local-vendor ()
+  #+(or linux windows) "-pc"
+  #+darwin "-apple"
+  #+(not (or linux windows darwin)) "-unknown")
+
+(defun local-os ()
+  #+linux "-linux"
+  #+windows "-windows-msvc"
+  #+darwin "-darwin9"
+  #+freebsd "-freebsd")
+
+(defun local-environment ()
+  #+linux "-gnu"
+  #-linux "")
+
+(defun local-arch ()
+  (strcat (local-cpu) (local-vendor) (local-os) (local-environment)))
+
+(defparameter *known-archs*
+  '("i686-pc-linux-gnu"
+    "x86_64-pc-linux-gnu"
+    "i686-pc-windows-msvc"
+    "x86_64-pc-windows-msvc"
+    "i686-apple-darwin9"
+    "x86_64-apple-darwin9"
+    "i386-unknown-freebsd"
+    "x86_64-unknown-freebsd"))
+
+(defvar *c2ffi-executable* "c2ffi")
+
+(defvar *trace-c2ffi* nil)
+
+(defun c2ffi-executable-available? ()
+  ;; This is a hack to determine if c2ffi exists; it assumes if it
+  ;; doesn't exist, we will get a return code other than 0.
+  (zerop (nth-value 2 (uiop:run-program `(,*c2ffi-executable* "-h") :ignore-error-status t))))
+
+(defun run-program* (program args &key (output (if *trace-c2ffi* *standard-output* nil))
+                                    (error-output (if *trace-c2ffi* *error-output* nil))
+                                    ignore-error-status)
+  (when *trace-c2ffi*
+    (format *debug-io* "~&; Invoking: ~A~{ ~A~}~%" program args))
+  (zerop (nth-value 2 (uiop:run-program (list* program args) :output output :error-output error-output :ignore-error-status ignore-error-status))))
+
+(defun generate-spec-with-c2ffi (input-header-file output-spec-path &key arch sys-include-paths ignore-error-status)
+  "Run c2ffi on `INPUT-HEADER-FILE`, outputting to `OUTPUT-FILE` and
+`MACRO-OUTPUT-FILE`, optionally specifying a target triple `ARCH`."
+  (uiop:with-temporary-file (:pathname tmp-macro-file
+                             :keep *trace-c2ffi*)
+    nil
+    :close-stream
+    (let* ((arch (when arch (list "-A" arch)))
+           (sys-include-paths (loop
+                                :for dir :in sys-include-paths
+                                :append (list "-i" dir))))
+      ;; Invoke c2ffi to emit macros into TMP-MACRO-FILE
```

Comment:
wrote a longer comment, i'll rebase and push -f.

      ;; Invoke c2ffi to first emit C #define's into TMP-MACRO-FILE. First ask c2ffi
      ;; to generate a file of C global variables that are assigned the value of
      ;; the #define's, so that in the second pass below the C compiler
      ;; evaluates for us their bodies and thus we can get hold of their value.
      ;; This is a kludge and eventually we could/should support generating
      ;; cffi-grovel files, and in grovel mode not rely on this kludge.


        ;; Write a tmp header file that #include's the original input file and
        ;; the above generated macros file which will form the input for our
        ;; final, second pass.